File: ios/NativeCore/Sources/NativeCore/TripMapProjection.swift
import Foundation

public struct MapMarker: Equatable, Identifiable {
    public enum Kind: Equatable { case saved, reserve, day(Int) }
    public let key: String
    public var id: String { key }
    public let coordinate: Coordinate
    public let label: String
    public let kind: Kind
    public let tripID: String?
    public let itemKey: ItemKey?
    public let favoriteID: String?
}
public struct MapRoute {
    public let day: Int
    public let markers: [MapMarker]
}
public struct MapProjection {
    public let markers: [MapMarker]
    public let unlocatedItemCount: Int
    public let routes: [MapRoute]
}
public enum TripMapProjection {
    public static func googlePlaceIDs(trip:TripDocument?,saved:[SavedPlaceDocument])->[String] {
        var items=saved.map(\.payload)
        if let trip { for section in [.reserve] + (1...trip.dayCount).map(ScheduleSection.day) { items += (try? TripSchedule.items(in:section,trip:trip)) ?? [] } }
        return Set(items.compactMap { item -> String? in
            guard item["nativePlaceSource"] as? String == "google",let id=item["placeId"] as? String,
                  (1...1024).contains(id.unicodeScalars.count) else { return nil }
            return id
        }).sorted()
    }
    public static func make(trip: TripDocument?, saved: [SavedPlaceDocument], resolved: [String: ResolvedPlace], now: Date = Date()) -> MapProjection {
        var markers = [MapMarker](), routes = [MapRoute](), missing = 0, seen = Set<String>()
        func coordinate(_ item: [String: Any]) -> Coordinate? {
            if item["nativePlaceSource"] as? String == "google" {
                guard let id = item["placeId"] as? String, let found = resolved[id], found.placeID == id,
                      (0..<300).contains(now.timeIntervalSince(found.fetchedAt)) else { return nil }
                return found.coordinate
            }
            return Coordinate.from(item: item)
        }
        func key(_ parts: [String]) -> String {
            // Encoding a tuple avoids collisions even when IDs contain separators.
            String(data: try! JSONSerialization.data(withJSONObject: parts, options: [.withoutEscapingSlashes]), encoding: .utf8)!
        }
        if let trip {
            for section in [.reserve] + (1...trip.dayCount).map(ScheduleSection.day) {
                var segment = [MapMarker]()
                func flush() {
                    if section.token > 0 && segment.count > 1 { routes.append(.init(day: section.token, markers: segment)) }
                    segment.removeAll()
                }
                for item in (try? TripSchedule.items(in: section, trip: trip)) ?? [] {
                    guard let rawID = item["id"], let id = try? ItemKey(json: rawID),
                          let point = coordinate(item) else { missing += 1; flush(); continue }
                    let markerKey = key(["trip", trip.id, id.token])
                    guard seen.insert(markerKey).inserted else { missing += 1; flush(); continue }
                    let marker = MapMarker(key: markerKey, coordinate: point, label: label(item), kind: section.token == 0 ? .reserve : .day(section.token), tripID: trip.id, itemKey: id, favoriteID: nil)
                    markers.append(marker); segment.append(marker)
                }
                flush()
            }
        }
        for row in saved {
            guard let point = coordinate(row.payload) else { missing += 1; continue }
            let markerKey = key(["saved",row.id])
            guard seen.insert(markerKey).inserted else { missing += 1; continue }
            markers.append(.init(key: markerKey, coordinate: point, label: label(row.payload), kind: .saved, tripID: nil, itemKey: nil, favoriteID: row.id))
        }
        return .init(markers: markers, unlocatedItemCount: missing, routes: routes)
    }
    private static func label(_ item: [String: Any]) -> String {
        let display = item["displayName"] as? String ?? ""
        return display.isEmpty ? item["name"] as? String ?? "이름 없는 장소" : display
    }
}
public struct MapBounds: Equatable {
    public let south, north, west, east, longitudeSpan: Double
    public static func fit(_ coordinates: [Coordinate]) -> MapBounds? {
        guard !coordinates.isEmpty else { return nil }
        let sorted = coordinates.map { $0.longitude < 0 ? $0.longitude + 360 : $0.longitude }.sorted()
        var gap = -1.0, index = 0
        for i in sorted.indices {
            let next = i + 1 < sorted.count ? sorted[i + 1] : sorted[0] + 360
            if next - sorted[i] > gap { gap = next - sorted[i]; index = i }
        }
        func signed(_ value: Double) -> Double { value > 180 ? value - 360 : value }
        return .init(south: coordinates.map(\.latitude).min()!, north: coordinates.map(\.latitude).max()!,
                     west: signed(sorted[(index + 1) % sorted.count]), east: signed(sorted[index]), longitudeSpan: 360 - gap)
    }
}

File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/domain/TripMapProjection.kt
package com.travelplaner.nativepreview.domain
import kotlinx.serialization.json.*

data class MapMarker(val key: String, val coordinate: Coordinate, val label: String, val kind: Kind, val tripID: String? = null, val itemKey: ItemKey? = null, val favoriteID: String? = null) {
    sealed interface Kind {
        data object Saved: Kind
        data object Reserve: Kind
        data class Day(val number: Int): Kind
    }
}
data class MapRoute(val day: Int,val markers: List<MapMarker>)
data class MapProjection(val markers: List<MapMarker>,val unlocatedItemCount: Int,val routes: List<MapRoute>)
object TripMapProjection {
    fun googlePlaceIDs(trip:TripDocument?,saved:List<SavedPlaceDocument>):List<String> {
        val items=saved.map { it.payload }.toMutableList()
        if(trip!=null) (listOf(ScheduleSection.Reserve)+(1..trip.dayCount).map { ScheduleSection.Day(it) }).forEach { items+=runCatching { TripSchedule.items(it,trip) }.getOrDefault(emptyList()) }
        return items.mapNotNull { item ->
            if((item["nativePlaceSource"] as? JsonPrimitive)?.content!="google") null
            else (item["placeId"] as? JsonPrimitive)?.takeIf { it.isString }?.content?.takeIf { it.codePointCount(0,it.length) in 1..1024 }
        }.distinct().sorted()
    }
    fun make(trip: TripDocument?,saved: List<SavedPlaceDocument>,resolved: Map<String,ResolvedPlace>,now: Long = System.currentTimeMillis()): MapProjection {
        val markers = mutableListOf<MapMarker>(); val routes = mutableListOf<MapRoute>(); var missing = 0
        val seen = mutableSetOf<String>()
        fun coordinate(item: JsonObject): Coordinate? {
            if ((item["nativePlaceSource"] as? JsonPrimitive)?.content == "google") {
                val id = (item["placeId"] as? JsonPrimitive)?.content ?: return null
                val found = resolved[id] ?: return null
                return found.coordinate.takeIf { found.placeID == id && now >= found.fetchedAt && now - found.fetchedAt < 300000 }
            }
            return Coordinate.from(item)
        }
        fun key(vararg parts: String) = JsonArray(parts.map(::JsonPrimitive)).toString()
        if (trip != null) {
            (listOf(ScheduleSection.Reserve) + (1..trip.dayCount).map { ScheduleSection.Day(it) }).forEach { section ->
                val segment = mutableListOf<MapMarker>()
                fun flush() { if (section.token > 0 && segment.size > 1) routes.add(MapRoute(section.token,segment.toList())); segment.clear() }
                (runCatching { TripSchedule.items(section,trip) }.getOrNull() ?: emptyList()).forEach { item ->
                    val id = runCatching { ItemKey.from(item.getValue("id")) }.getOrNull()
                    val point = coordinate(item)
                    if (id == null || point == null) { missing++; flush() }
                    else {
                        val markerKey = key("trip",trip.id,id.token)
                        if (!seen.add(markerKey)) { missing++; flush() }
                        else {
                            val marker = MapMarker(markerKey,point,label(item),if (section.token == 0) MapMarker.Kind.Reserve else MapMarker.Kind.Day(section.token),trip.id,id)
                            markers.add(marker); segment.add(marker)
                        }
                    }
                }
                flush()
            }
        }
        saved.forEach { row ->
            val point = coordinate(row.payload); val markerKey = key("saved",row.id)
            if (point == null || !seen.add(markerKey)) missing++
            else markers.add(MapMarker(markerKey,point,label(row.payload),MapMarker.Kind.Saved,favoriteID=row.id))
        }
        return MapProjection(markers.toList(),missing,routes.toList())
    }
    private fun label(item: JsonObject): String = (item["displayName"] as? JsonPrimitive)?.content?.takeIf { it.isNotEmpty() }
        ?: (item["name"] as? JsonPrimitive)?.content ?: "이름 없는 장소"
}
data class MapBounds(val south: Double,val north: Double,val west: Double,val east: Double,val longitudeSpan: Double) {
    companion object {
        fun fit(coordinates: List<Coordinate>): MapBounds? {
            if (coordinates.isEmpty()) return null
            val sorted = coordinates.map { if (it.longitude < 0) it.longitude + 360 else it.longitude }.sorted()
            var gap = -1.0; var index = 0
            sorted.indices.forEach { i ->
                val next = if (i+1 < sorted.size) sorted[i+1] else sorted[0]+360
                if (next-sorted[i] > gap) { gap = next-sorted[i]; index=i }
            }
            fun signed(value: Double) = if (value > 180) value-360 else value
            return MapBounds(coordinates.minOf { it.latitude },coordinates.maxOf { it.latitude },signed(sorted[(index+1)%sorted.size]),signed(sorted[index]),360-gap)
        }
    }
}
